ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,252 @@
+
+    for (Field field : dataSource.getSchema().fields()) {
+      if (!field.schema().isOptional() && values.getOrDefault(field.name(), null) == null) {
+        throw new KsqlException("Got null value for nonnull field: " + field);
+      }
+    }
+
+    return new GenericRow(
+        dataSource.getSchema()
+            .fields()
+            .stream()
+            .map(Field::name)
+            .map(values::get)
+            .collect(Collectors.toList())
+    );
+  }
+
+  @SuppressWarnings("unchecked") // we know that key is String
+  private byte[] serializeKey(final Object keyValue, final StructuredDataSource<?> dataSource) {
+    if (keyValue == null) {
+      return null;
+    }
+
+    return ((Serde<String>) dataSource.getKeySerdeFactory().create())
+        .serializer()
+        .serialize(dataSource.getKafkaTopicName(), keyValue.toString());
+  }
+
+  private byte[] serializeRow(
+      final GenericRow row,
+      final StructuredDataSource<?> dataSource,
+      final KsqlConfig config,
+      final ServiceContext serviceContext
+  ) {
+    final Serde<GenericRow> rowSerde = dataSource.getKsqlTopicSerde()
+        .getGenericRowSerde(
+            dataSource.getSchema(),
+            config,
+            serviceContext.getSchemaRegistryClientFactory(),
+            "",
+            NoopProcessingLogContext.INSTANCE);
+    return rowSerde.serializer().serialize(dataSource.getKafkaTopicName(), row);
+  }
+
+  private static class ExpressionResolver extends AstVisitor<Object, Void> {
+
+    private static final Map<Type, Function<Number, Object>> CASTER =
+        ImmutableMap.<Type, Function<Number, Object>>builder()
+            .put(Type.INT8, Number::shortValue)
+            .put(Type.INT32, Number::intValue)
+            .put(Type.INT64, Number::longValue)
+            .put(Type.FLOAT32, Number::floatValue)
+            .put(Type.FLOAT64, Number::doubleValue)
+            .build();
+
+    private final Schema schema;
+    private final String field;
+
+    ExpressionResolver(final Schema schema, final String field) {
+      this.schema = schema;
+      this.field = field;
+    }
+
+    @Override
+    protected String visitNode(final Node node, final Void context) {
+      throw new KsqlException(
+          "Only Literals are supported for INSERT INTO. Got: " + node + " for field " + field);
+    }
+
+    @Override
+    protected Object visitLiteral(final Literal node, final Void context) {
+      final Object value = node.getValue();
+      if (node instanceof NullLiteral || value == null) {
+        return null;
+      }
+
+      final Type valueType = SchemaUtil.getSchemaFromType(value.getClass()).type();
+      if (valueType.equals(schema.type())) {
+        return value;
+      }
+
+      if (SchemaUtil.isNumber(valueType)
+          && SchemaUtil.isNumber(schema.type())
+          && SchemaUtil.canUpCast(schema.type(), valueType)) {
+        return CASTER.get(schema.type()).apply((Number) value);
```

Comment:
As above, I'm not convinced casting from double to int is something we want to do implicitly.

Reply:
See above, this wouldn't happen but I'll remove it anyway.